fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogFragment.kt

```diff
@@ -128,12 +171,14 @@ class TranslationsDialogFragment : BottomSheetDialogFragment() {
                                     onTranslateButtonClick = {
                                         interactor.onTranslate(
                                             tabId = args.sessionId,
-                                            fromLanguage = null,
-                                            toLanguage = null,
+                                            fromLanguage = fromSelected?.code,
```

Comment:
Filed [Bug 1875534](https://bugzilla.mozilla.org/show_bug.cgi?id=1875534) to follow-up on the condition when the from and to language are the same.

Reply:
In what class do we do all these validations? This will not be the only one needed in this fragment. @gabrielluong

Reply:
If a language is selected, do we still show it in the drop-down menu? I asked in figma. https://www.figma.com/file/GT6oZLFw32LiCwKl5GOMWD?node-id=7202:119619&mode=design#677742095

Reply:
In what class do we modify the state of the views? For example, if the translation is successful, this fragment should look different. The first dropdown should disappear, the title should be different, and so should the buttons. 
Where do we put this logic ?

Reply:
> In what class do we do all these validations?  
> In what class do we modify the state of the views?

These are the big question right now about the overall architecture this project should be heading towards. We should take some time to formally write out a plan or diagram because there is so much involved. It seems especially tricky where almost everything is happening on the GeckoView engine layer and reporting on the browser store, so we need a pattern that supports that data model. Maybe we should try and do a Figma Jam or something together with anyone interested? 

> In what class do we modify the state of the views?

My super naive thought is: can we just observe everything in the fragment and update based on that? For example, I had tested around some with having  the `translationError`  be observed and set on `TranslationDialogBottomSheet`. It showed the error as I expected and seemed to be managing what I expected. 

I imagine this type of logic might become more profound with all of the additional criteria, but in general, the `browserStore` observed values seems to correlate fairly strongly to the UI needs.

I have a sense there is a reason not too do what I'm thinking out loud about and use formal `States`, but don't really understand the why aspect and would like to know more.

> In what class do we do all these validations?  

Would it be acceptable to do validation in the fragment as well when it is view related? E.g., observe the `from` and `to`, if they are the same then disable the button. I suppose the logic might get overwhelming.

Mostly want to get a sense of why certain things aren't a good idea too so we can start focusing in on a pattern.

Also, worth mentioning, the goal of this bug is only to give us a few more pieces lined out to work with as a starting point, so I expect whatever lands here will be refactored or change as we decide on a pattern.

Reply:
This is a preliminary comment to more architectural advice (if desired), because I don't have a ton of time today but would love to talk more about all this. Have only done a cursory reading of some of the material so far. Here are some suggestions I would make to answer the following questions and in general for this feature to keep it in line with our Redux architectural goals:

The workflow for Redux is:
1. Dispatch an action
2. Handle side-effects and logic in middleware
3. Update state in a reducer
4. Observe updates to that state in the view and update accordingly

So, to answer the above:

> In what class do we do all these validations?

These can happen in the reducer (if they are pure computations) or in the middleware (if they are not, for example they require accessing disk state through `Settings`). In the case of middleware, this may result in the original action being stopped from entering the rest of reducer chain, though that does add some complexity.

> In what class do we modify the state of the views?

The views should modify themselves as they observe changes to `State`. I would recommend having State that isn't specific to a view, for example it would contain a `translationError` instead of a `translationErrorString`. 

> I have a sense there is a reason not too do what I'm thinking out loud about and use formal States, but don't really understand the why aspect and would like to know more.

I think I'd need a bit more detail here to comment either way. 

> E.g., observe the from and to, if they are the same then disable the button.

This is an interesting case. I'll think about this before I can answer definitively, but some questions we can ask about it:
1. Do we want to be able to unit test the fact the button is disabled? If so, does it make sense to keep it in `State`, or will we be computing it every time we want to inspect it? Could we expose the logic through something like an extension function and test that easily?

Finally, I see that the interactor/controller have landed. My suggestion would be to remove those completely, along with the `TranslateUseCase`. Instead, I would dispatch Actions directly from the callback passed to the view (like `onTranslateButtonClick` referenced in the diff of this comment). This is because the current workflow is: callback -> interactor -> controller -> usecase -> dispatch action. I don't see any reason why we can't remove the interactor -> controller -> usecase middlemen here.

Reply:
> Filed Bug 1875534 to follow-up on the condition when the from and to language are the same.
> In what class do we do all these validations? This will not be the only one needed in this fragment.

IMO this is something that the view can handle directly, and is much easier in Compose. Something like:

```
Button(isEnabled = state.fromLanguage != state.ToLanguage)
```

is not necessarily about stressing unit tests for. 

> In what class do we modify the state of the views? For example, if the translation is successful, this fragment should look different. The first dropdown should disappear, the title should be different, and so should the buttons.

The views should control their own state in response to observations of feature state. In all of these cases, I would not suggest introducing another class to hold view state.

Reply:
>If a language is selected, do we still show it in the drop-down menu? 

Closing the loop here, answer was yes the language should be in the drop-down menu, but the design changed to included a checkmark next to the selected item. Filed new UI bug - [Bug 1876140](https://bugzilla.mozilla.org/show_bug.cgi?id=1876140)